File: lib/languages/zig/type-checking.zig
// Type checking examples
const name: []const u8 = "Alice";
// const name: []const u8 = 42;  // Compile error

// Function with type annotations
fn process(value: i32) []const u8 {
    return std.fmt.allocPrint(allocator, "{d}", .{value}) catch unreachable;
}

// Calling with wrong type
// process("hello");  // Compile error

// Comptime type checking
fn min(comptime T: type, a: T, b: T) T {
    return if (a < b) a else b;
}

// Using the generic function
const int_min = min(i32, 5, 10);
const float_min = min(f64, 5.5, 10.1);

// Type coercion is limited
const x: f64 = 42.0;
// const y: i32 = x;  // Error: no implicit conversion

// Explicit conversion
const z: i32 = @floatToInt(i32, x);

// Type inference
const inferred = 42; // i32 is inferred
var array = [_]i32{1, 2, 3}; // [3]i32 is inferred

File: lib/languages/rust/type-checking.rs
// Type checking examples
let name: String = String::from("Alice");
// let name: String = 42;  // Compile error

// Function with type annotations
fn process(value: i32) -> String {
   value.to_string()
}

// Calling with wrong type
// process("hello");  // Compile error

// Type inference
let inferred = 42; // i32 is inferred
let vector = vec![1, 2, 3]; // Vec<i32> is inferred

// Type coercion is limited
let x: f64 = 42.0;
// let y: i32 = x;  // Error: no implicit conversion

// Explicit conversion
let z: i32 = x as i32;

